File: scripts/scrapers/louvre.py
"""Scraper du musée du Louvre.

La page « expositions » est un site Next.js : les expositions sont des objets
`{"type": "Exposition", ...}` dans le JSON __NEXT_DATA__. Pas de HTML à parser.

On ne garde que les expositions EN COURS et physiquement au Louvre (on exclut
les expositions passées, à venir, et « Le Louvre ailleurs » = hors les murs).
Pour chaque expo retenue, on suit sa fiche pour récupérer la description longue
(≈ 1 requête de plus par expo, soit une poignée).
"""
import re

from . import base

SITE = "https://www.louvre.fr"
IMG_HOST = "https://api-www.louvre.fr"   # les médias sont servis par l'hôte d'API


def _image(img):
    """URL d'image : variante paysage 3:2 optimisée (webp), sinon l'original.

    Les images du Louvre sont servies par api-www.louvre.fr (pas www), et le
    JSON fournit des dérivés prêts à l'emploi dans `hashes`.
    """
    if not isinstance(img, dict):
        return None
    hashes = img.get("hashes") or {}
    for w in (1200, 1080, 828, 750, 640):
        url = hashes.get(f"w{w}_3_2")
        if url:
            return url
    path = img.get("fallback") or img.get("path")
    return base.absolute(IMG_HOST, path) if path else None


def _walk(node, wanted_type=None, key=None):
    """Parcourt récursivement et rend les dicts ayant type==wanted_type,
    ou les valeurs str trouvées à la clé `key`."""
    if isinstance(node, dict):
        if wanted_type and node.get("type") == wanted_type:
            yield node
        for k, v in node.items():
            if key and k == key and isinstance(v, str):
                yield v
            yield from _walk(v, wanted_type, key)
    elif isinstance(node, list):
        for v in node:
            yield from _walk(v, wanted_type, key)


def _is_current_onsite(tags):
    low = [t.lower() for t in tags]
    if not any(t.startswith("en cours") for t in low):
        return False
    if any("ailleurs" in t for t in low):   # « Le Louvre ailleurs » = hors les murs
        return False
    return True


def _fetch_description(url):
    """Récupère la description longue depuis la fiche détaillée."""
    from paname.utils import html_to_text
    try:
        data = base.next_data(base.fetch(url))
    except Exception:
        return None
    if not data:
        return None
    page = data.get("props", {}).get("initialState", {}).get("Page", {})
    # Tous les blocs riches sont des `...html`. On prend le plus long « vrai »
    # paragraphe (sans gabarits {{ ... }} de crédits / infos pratiques).
    candidates = []
    for raw in _walk(page, key="html"):
        if "{{" in raw:
            continue
        clean = re.sub(r"<\?xml[^>]*\?>", "", raw)
        txt = html_to_text(clean)
        if txt:
            candidates.append(txt)
    return max(candidates, key=len) if candidates else None


def scrape(museum):
    data = base.next_data(base.fetch(museum.expos_url))
    if not data:
        return []

    items = []
    for e in _walk(data, wanted_type="Exposition"):
        title = (e.get("title") or "").strip()
        link = e.get("link") or {}
        url_path = link.get("url") if isinstance(link, dict) else link
        if not title or not url_path:
            continue

        tags = [t.get("label", "").strip() for t in e.get("tags", []) if isinstance(t, dict)]
        if not _is_current_onsite(tags):
            continue

        img = e.get("image") or {}
        date_start, date_end = base.parse_french_date(e.get("date"))
        url = base.absolute(SITE, url_path)
        subtitle = (e.get("subtitle") or "").strip() or None

        items.append({
            "title": title,
            "subtitle": subtitle,
            "description": _fetch_description(url) or subtitle,
            "url": url,
            "image_url": _image(img),
            "date_start": date_start,
            "date_end": date_end,
            "tags": tags,
            "ext_key": url_path.rstrip("/").rsplit("/", 1)[-1],
        })
    return items

File: src/scrapers/louvre.js
"use strict";

// Scraper du musée du Louvre (site Next.js : expositions dans __NEXT_DATA__).
// On ne garde que les expositions EN COURS et physiquement au Louvre.
// Port de scripts/scrapers/louvre.py.

const base = require("./base");
const { htmlToText } = require("../lib/text");

const SITE = "https://www.louvre.fr";
const IMG_HOST = "https://api-www.louvre.fr";

function image(img) {
  if (!img || typeof img !== "object") return null;
  const hashes = img.hashes || {};
  for (const w of [1200, 1080, 828, 750, 640]) {
    const url = hashes[`w${w}_3_2`];
    if (url) return url;
  }
  const path = img.fallback || img.path;
  return path ? base.absolute(IMG_HOST, path) : null;
}

// Parcourt récursivement, collecte les objets {type == wantedType}.
function walkType(node, type, out = []) {
  if (Array.isArray(node)) {
    for (const v of node) walkType(v, type, out);
  } else if (node && typeof node === "object") {
    if (node.type === type) out.push(node);
    for (const v of Object.values(node)) walkType(v, type, out);
  }
  return out;
}

// Collecte les valeurs string trouvées à la clé `key`.
function walkKey(node, key, out = []) {
  if (Array.isArray(node)) {
    for (const v of node) walkKey(v, key, out);
  } else if (node && typeof node === "object") {
    for (const [k, v] of Object.entries(node)) {
      if (k === key && typeof v === "string") out.push(v);
      walkKey(v, key, out);
    }
  }
  return out;
}

function isCurrentOnsite(tags) {
  const low = tags.map((t) => t.toLowerCase());
  if (!low.some((t) => t.startsWith("en cours"))) return false;
  if (low.some((t) => t.includes("ailleurs"))) return false; // hors les murs
  return true;
}

async function fetchDescription(url) {
  let data;
  try {
    data = base.nextData(await base.fetchText(url));
  } catch (e) {
    return null;
  }
  if (!data) return null;
  const page = (((data.props || {}).initialState || {}).Page) || {};
  const candidates = [];
  for (const raw of walkKey(page, "html")) {
    if (raw.includes("{{")) continue;
    const clean = raw.replace(/<\?xml[^>]*\?>/g, "");
    const txt = htmlToText(clean);
    if (txt) candidates.push(txt);
  }
  return candidates.length ? candidates.reduce((a, b) => (b.length > a.length ? b : a)) : null;
}

async function scrape(museum) {
  const data = base.nextData(await base.fetchText(museum.expos_url));
  if (!data) return [];

  const items = [];
  for (const e of walkType(data, "Exposition")) {
    const title = (e.title || "").trim();
    const link = e.link || {};
    const urlPath = typeof link === "object" ? link.url : link;
    if (!title || !urlPath) continue;

    const tags = (e.tags || [])
      .filter((t) => t && typeof t === "object")
      .map((t) => (t.label || "").trim());
    if (!isCurrentOnsite(tags)) continue;

    const [dateStart, dateEnd] = base.parseFrenchDate(e.date);
    const url = base.absolute(SITE, urlPath);
    const subtitle = (e.subtitle || "").trim() || null;

    items.push({
      title,
      subtitle,
      description: (await fetchDescription(url)) || subtitle,
      url,
      image_url: image(e.image || {}),
      date_start: dateStart,
      date_end: dateEnd,
      tags,
      ext_key: urlPath.replace(/\/+$/, "").split("/").pop(),
    });
  }
  return items;
}

module.exports = { scrape };
